# Минипроект 1

### Описание данных

Имеются следующие переменные:

Hotel – тип отеля (City Hotel или Resort Hotel)  

Is canceled – бронирование было отменено (1) или нет (0); неотменённое считается успешным

Lead time – количество дней, прошедших между датой бронирования и датой прибытия  

Arrival full date – полная дата прибытия

Arrival date year – год прибытия  

Arrival date month – месяц прибытия  

Arrival date week number – номер недели прибытия

Arrival date day of month – день прибытия

Stays in weekend nights – количество выходных (суббота или воскресенье), которые гость забронировал для проживания в отеле

Stays in week nights – количество дней (с понедельника по пятницу), которые гость забронировал для проживания в отеле

Stays total nights – общее число забронированных ночей (сумма двух предыдущих колонок)

Adults – число взрослых

Children – число детей

Babies – число младенцев 

Meal – выбранный тип питания

Country – страна происхождения клиента

Reserved room type – тип зарезервированного номера

Assigned room type – тип полученного номера (может отличаться от забронированного)

Customer type – тип бронирования

Reservation status – значение последнего статуса брони: Canceled – было отменено клиентом; Check-Out – клиент зарегистрировался, но уже покинул отель; No-Show – клиент не зарегистрировался и сообщил администрации отеля причину

Reservation status date – дата обновления статуса

In [3]:
import pandas as pd

In [8]:
bookings = pd.read_csv('https://stepik.org/media/attachments/lesson/360344/bookings.csv', sep = ';')

In [10]:
bookings.head(10)

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   
5  Resort Hotel            0         14        2015-07-01               2015   
6  Resort Hotel            0          0        2015-07-01               2015   
7  Resort Hotel            0          9        2015-07-01               2015   
8  Resort Hotel            1         85        2015-07-01               2015   
9  Resort Hotel            1         75        2015-07-01               2015   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   
5               July                        27                          1   
6               July                        27                          1   
7               July                        27                          1   
8               July                        27                          1   
9               July                        27                          1   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   
5                        0                     2  ...       2       0.0   
6                        0                     2  ...       2       0.0   
7                        0                     2  ...       2       0.0   
8                        0                     3  ...       2       0.0   
9                        0                     3  ...       2       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   
5       0    BB     GBR                  A                  A     Transient   
6       0    BB     PRT                  C                  C     Transient   
7       0    FB     PRT                  C                  C     Transient   
8       0    BB     PRT                  A                  A     Transient   
9       0    HB     PRT                  D                  D     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  
5          Check-Out              2015-07-03  
6          Check-Out              2015-07-03  
7          Check-Out              2015-07-03  
8           Canceled              2015-05-06  
9           Canceled              2015-04-22  

[10 rows x 21 columns]

### Задание 2
Посмотрите на размеры таблицы. Сколько всего в ней столбцов?

In [4]:
bookings.shape[1]

21

### Задание 3
К какому типу относится большинство переменных? 

In [17]:
bookings.dtypes.value_counts()

object     10
int64      10
float64     1
dtype: int64

Ответ: object, int64

### Задание 4
Как вы уже могли заметить, при сборе датасета произошла небольшая ошибка, и названия колонок были записаны в неудобном для дальнейшей работы формате. Исправьте ситуацию: приведите названия столбцов к нижнему регистру и замените пробелы на знак нижнего подчёркивания. Например: Is Canceled -> is_canceled

Могут пригодиться: 

rename

str.lower

str.replace

columns

In [30]:
def column_rename(name):
    return name.replace(' ', '_').lower()

bookings = bookings.rename(columns = column_rename)

In [32]:
bookings.head()

hotel  is_canceled  lead_time arrival_full_date  arrival_date_year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   

  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  ...  adults  children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   

   babies  meal country reserved_room_type assigned_room_type customer_type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   

  reservation_status reservation_status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  

[5 rows x 21 columns]

### Задание 5
Перейдём к исследованию данных! Пользователи из каких стран совершили наибольшее число успешных бронирований? 
Бронирование считается успешным, если в дальнейшем оно не было отменено (переменная is_canceled). В качестве ответа выберите страны, входящие в топ-5.

In [39]:
bookings.query("is_canceled == 0").\
        groupby('country', as_index = False).\
        agg({'is_canceled': 'count'}).\
        sort_values('is_canceled', ascending = False).\
        head(5)

country  is_canceled
125     PRT        21071
57      GBR         9676
54      FRA         8481
50      ESP         6391
42      DEU         6069

### Задание 6
На сколько ночей (stays_total_nights)  в среднем бронируют отели типа City Hotel? Resort Hotel? Запишите полученные значения с точностью до 2 знаков после точки.


Может пригодиться: round

In [46]:
bookings.groupby('hotel')\
        .agg({'stays_total_nights': 'mean'})\
        .round(2)

stays_total_nights
hotel                           
City Hotel                  2.98
Resort Hotel                4.32

### Задание 7
Иногда тип номера, полученного клиентом (assigned_room_type), отличается от изначально забронированного (reserved_room_type). Такое может произойти, например, по причине овербукинга.

Сколько подобных наблюдений встретилось в датасете?

*отмена бронирования также считается

In [50]:
bookings.query("assigned_room_type != reserved_room_type").shape[0]

14917

### Задание 8
Теперь проанализируйте даты запланированного прибытия (arrival_date_year).

На какой месяц чаще всего оформляли бронь в 2016 году? Изменился ли самый популярный месяц в 2017 году?

In [56]:
bookings.query('arrival_date_year == 2016') \
        .arrival_date_month \
        .value_counts()

October      6203
May          5478
April        5428
September    5394
June         5292
August       5063
March        4824
July         4572
November     4454
February     3891
December     3860
January      2248
Name: arrival_date_month, dtype: int64

In [57]:
bookings.query('arrival_date_year == 2017') \
        .arrival_date_month \
        .value_counts()

May         6313
April       5661
June        5647
July        5313
March       4970
August      4925
February    4177
January     3681
Name: arrival_date_month, dtype: int64

В 2016 году чаще всего бронь оформляли на октябрь. В 2017 ситуация изменилась. Самым популярным месяцем стал май.

### Задание 9 
Сгруппируйте данные по годам, а затем проверьте, на какой месяц (arrival_date_month) бронирования отеля типа City Hotel отменялись чаще всего в 2015? 2016? 2017? 

Hint: для этого сначала примените метод groupby, после чего выберите нужную колонку для дальнейших действий :) Например:

df.groupby('column_1')['column_2'].what_to_do()

In [66]:
bookings.query('is_canceled == 1 and hotel == "City Hotel"') \
        .groupby(['arrival_date_year', 'arrival_date_month']) \
        .agg({'is_canceled': 'count'}) \
        .sort_values(['arrival_date_year', 'is_canceled'], ascending = False)

is_canceled
arrival_date_year arrival_date_month             
2017              May                        2217
                  April                      1926
                  June                       1808
                  July                       1324
                  March                      1278
                  August                     1123
                  January                    1044
                  February                    971
2016              October                    1947
                  June                       1720
                  September                  1567
                  April                      1539
                  May                        1436
                  November                   1360
                  August                     1247
                  March                      1108
                  December                   1072
                  July                       1043
                  February                    930
                  January                     438
2015              September                  1543
                  October                    1321
                  August                     1232
                  July                        939
                  December                    668
                  November                    301

2015: сентябрь

2016: октябрь

2017: май

### Задание 10
Посмотрите на числовые характеристики трёх колонок: adults, children и babies. Какая из них имеет наибольшее среднее значение?

In [67]:
bookings.adults.mean()

1.8564033838679956

In [68]:
bookings.children.mean()

0.10388990333874994

In [69]:
bookings.babies.mean()

0.007948739425412514

Наибольшее значение - adults

### Задание 11
Создайте колонку total_kids, объединив столбцы children и babies. Для отелей какого типа среднее значение переменной оказалось наибольшим?

City hotel – отель находится в городе
Resort hotel – отель курортный
В качестве ответа укажите наибольшее среднее total_kids, округлив до 2 знаков после точки.

In [72]:
#def sum_columns(col1, col2):
#    return col1 + col2

In [73]:
#bookings['total_kids'] = sum_columns(bookings.children, bookings.babies)

In [79]:
#del bookings['total_kids']

In [80]:
bookings['total_kids'] = bookings.children + bookings.babies

In [83]:
bookings.groupby('hotel') \
        .agg({'total_kids': 'mean'}) \
        .round(2)

total_kids
hotel                   
City Hotel          0.10
Resort Hotel        0.14

Ответ: 0.14

### Задание 12
Не все бронирования завершились успешно (is_canceled), поэтому можно посчитать, сколько клиентов было потеряно в процессе. Иными словами, посчитать метрику под названием Churn Rate.

Churn rate (отток, коэффициент оттока) – это процент подписчиков (например, на push-уведомления от сайта), которые отписались от канала коммуникации, отказались от услуг сервиса в течение определенного периода времени. Иными словами, представляет собой отношение количества ушедших пользователей к общему количеству пользователей, выраженное в процентах.

В нашем случае Churn Rate - это процент клиентов, которые отменили бронирование. Давайте посмотрим, как эта метрика связана с наличием детей у клиентов!

Создайте переменную has_kids, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (total_kids), в противном случае – False. Далее проверьте, среди какой группы пользователей показатель оттока выше. 

В качестве ответа укажите наибольший %, округленный до 2 знаков после точки (то есть доля 0.24563 будет 24.56% и в ответ пойдёт 24.56)



In [90]:
bookings['has_kids'] = bookings.total_kids > 0

In [91]:
bookings['has_kids']

0         False
1         False
2         False
3         False
4         False
          ...  
119385    False
119386    False
119387    False
119388    False
119389    False
Name: has_kids, Length: 119390, dtype: bool

In [95]:
round(((bookings.query('is_canceled == 1 and has_kids == False').shape[0] / bookings.query('has_kids == False').shape[0]) * 100), 2)

37.22

In [96]:
round(((bookings.query('is_canceled == 1 and has_kids == True').shape[0] / bookings.query('has_kids == True').shape[0]) * 100), 2)

34.92

Показатель оттока выше среди людей, у которых нет детей и равен 37.22%